src/plugins/direct_access_links/public/plugin.ts

```diff
@@ -0,0 +1,85 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { PluginInitializerContext, CoreSetup, CoreStart, Plugin } from 'src/core/public';
+import { i18n } from '@kbn/i18n';
+import {
+  AccessLinkGenerator,
+  GeneratorId,
+  DirectAccessLinkOptions,
+  createDirectAccessLinkGenerator,
+} from './direct_access_link_generator';
+
+export interface DirectAccessLinksStart {
+  getAccessLinkGenerator: (urlGeneratorId: GeneratorId) => AccessLinkGenerator<GeneratorId>;
+}
+
+export interface DirectAccessLinksSetup {
+  registerAccessLinkGenerator: <Id extends GeneratorId>(
+    id: Id,
+    generator: DirectAccessLinkOptions<Id>
+  ) => void;
+}
+
+export class DirectAccessLinksPlugin
+  implements Plugin<DirectAccessLinksSetup, DirectAccessLinksStart> {
+  private accessLinkGenerators: {
+    [Id in GeneratorId]?: AccessLinkGenerator<Id>;
+  } = {};
+
+  constructor(initializerContext: PluginInitializerContext) {}
+
+  public setup(core: CoreSetup) {
+    const setup: DirectAccessLinksSetup = {
+      registerAccessLinkGenerator: <Id extends GeneratorId>(
+        id: Id,
+        generatorOptions: DirectAccessLinkOptions<Id>
+      ) => {
+        this.accessLinkGenerators[id] = createDirectAccessLinkGenerator(
+          generatorOptions,
+          this.getAccessLinkGenerator.bind(this)
+        );
+      },
+    };
+    return setup;
+  }
+
+  public start(core: CoreStart) {
+    const start: DirectAccessLinksStart = {
+      getAccessLinkGenerator: this.getAccessLinkGenerator.bind(this),
+    };
+    return start;
+  }
+
+  public stop() {}
+
+  private getAccessLinkGenerator(id: GeneratorId) {
```

Comment:
nit, if you define this function as a class property, you will not need to `.bind()` it in *setup* and *start* life-cycles.

```suggestion
  private readonly getAccessLinkGenerator = (id: GeneratorId) => {
```